infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -73,6 +100,140 @@ public boolean isPredefined() {
+    // the default namespace must be configured if user defined are not allowed
+    // so return only it
+    String evaluatedName =
+        evalDefaultNamespaceName(
+            defaultNamespaceName, EnvironmentContext.getCurrent().getSubject());
+
+    Optional<KubernetesNamespaceMeta> defaultNamespaceOpt = fetchNamespace(evaluatedName);
+
+    KubernetesNamespaceMeta defaultNamespace =
+        defaultNamespaceOpt
+            // if the predefined namespace does not exist - return dummy info and it will be created
+            // during the first workspace start
+            .orElseGet(() -> new KubernetesNamespaceMetaImpl(evaluatedName));
+
+    defaultNamespace.getAttributes().put(DEFAULT_ATTRIBUTE, "true");
+    return defaultNamespace;
+  }
+
+  /**
+   * Provision default namespace into the specified list. If default namespace is already there -
+   * just provision the corresponding attributes to it.
+   *
+   * @param namespaces list where default namespace should be provisioned
+   */
+  private void provisionDefaultNamespace(List<KubernetesNamespaceMeta> namespaces) {
+    String evaluatedName =
+        evalDefaultNamespaceName(
+            defaultNamespaceName, EnvironmentContext.getCurrent().getSubject());
+
+    Optional<KubernetesNamespaceMeta> defaultNamespaceOpt =
+        namespaces.stream().filter(n -> evaluatedName.equals(n.getName())).findAny();
+    KubernetesNamespaceMeta defaultNamespace;
+    if (defaultNamespaceOpt.isPresent()) {
+      defaultNamespace = defaultNamespaceOpt.get();
+    } else {
+      defaultNamespace = new KubernetesNamespaceMetaImpl(evaluatedName);
+      namespaces.add(defaultNamespace);
+    }
+
+    defaultNamespace.getAttributes().put(DEFAULT_ATTRIBUTE, "true");
+  }
+
+  /**
+   * Fetches the specified namespace from a cluster.
+   *
+   * @param name name of namespace that should be fetched.
+   * @return optional with kubernetes namespace meta
+   * @throws InfrastructureException when any error occurs during namespace fetching
+   */
+  protected Optional<KubernetesNamespaceMeta> fetchNamespace(String name)
+      throws InfrastructureException {
+    try {
+      Namespace namespace = clientFactory.create().namespaces().withName(name).get();
+      if (namespace == null) {
+        return Optional.empty();
+      } else {
+        return Optional.of(asNamespaceMeta(namespace));
+      }
+    } catch (KubernetesClientException e) {
+      throw new InfrastructureException(
+          "Error occurred when tried to fetch default namespace. Cause: " + e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Fetched namespace from a k8s cluster.
+   *
+   * @return list with available k8s namespace metas.
+   * @throws InfrastructureException when any error occurs during namespaces fetching
+   */
+  protected List<KubernetesNamespaceMeta> fetchNamespaces()
+      throws KubernetesClientException, InfrastructureException {
```

Comment:
Looks like `KubernetesClientException` cannot be thrown here

Reply:
refactoring issue. fixed